ORC-742: LazyIO for non-filter columns

pavibhai opened the pull request:
### What changes were proposed in this pull request?
Added Lazy IO for follow columns.
        * Identify columns in the presence of a filter into LEAD and FOLLOW columns
                * LEAD columns are read first
                * FOLLOW columns are read only if the filter selects an output
                * This evaluation is reset on every stripe change
        * RecordReaderImpl.nextBatch performs read until a batch has value or the file is exhaused instead of returning empty batches as was the case previously
        * IO of FOLLOW columns happens the same as partial RowGroup selections during read
        * In the presence of filters batches respected row group boundaries
        * Filter is now defined as Consumer<FilterContext> instead of Consumer<VectorizedRowBatch>

### Why are the changes needed?
* The code changes allow for a lazy evaluation of FOLLOW columns, which in the case of reads with minimal hits gives substantial savings both of IO and CPU.
* The filter is changed to Consumer<FilterContext> to offer a convenience method on retrieving a ColumnVector using a name `FilterContext.findVector`


### How was this patch tested?
* This patch includes Unit tests that verify the IO savings accomplished as a result of this change.
* Given the interface and behavior change of the filters, some of the existing unit tests were updated to reflect the new API as well as new behavior of not reading FOLLOW columns unless required.

dongjoon-hyun commented:
Could your resolve conflicts by rebasing to the master branch?

pavibhai commented:
> Could your resolve conflicts by rebasing to the master branch?

done

dongjoon-hyun commented on .gitignore:
```diff
@@ -9,3 +9,4 @@ target
 *.iws
 .idea
 .DS_Store
+*.monopic
```
I'm wondering if this is required in this PR.

pavibhai replied:
Removed

dongjoon-hyun commented on java/core/src/java/org/apache/orc/Reader.java:
```diff
@@ -265,7 +269,7 @@ public Options schema(TypeDescription schema) {
      * @return this
      */
-    public Options setRowFilter(String[] filterColumnNames, Consumer<VectorizedRowBatch> filterCallback) {
+    public Options setRowFilter(String[] filterColumnNames, Consumer<FilterContext> filterCallback) {
```
This looks like a breaking API change.

pavibhai replied:
Yes this is a breaking change. I thought the convenience of retrieving the column vectors via name does ease the complexity of filter development.

I am open to other ideas on achieving the same.

dongjoon-hyun replied:
As you know, we cannot change like this. You may try to add as an additional new function instead.

pavibhai replied:
This is not yet released in ORC so I was expecting that we should be flexible on changing this.

pavibhai replied:
Merged into master as part of ORC-755

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/BufferChunk.java:
```diff
@@ -69,22 +69,8 @@ public final String toString() {
-    }
-    return new BufferChunk(sliceBuf, offset + shiftBy);
+    // Method has no code references
+    throw new UnsupportedOperationException();
```
If this method is not used, we need to to in an independent PR.
Please revert this change from this PR.

dongjoon-hyun replied:
Of course, we should evaluate the removal PR first.
For that PR, we need to validate with ORC/Iceberg/Hive at least.
Although there exists many downstream, currently only the above three are compatible with ORC 1.6.x.
We don't want to lose them again.

pavibhai replied:
I have gone through the change again carefully. This is not required for the current implementation, so I will revert this change here and we can take this up independently if needed

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/BufferChunkList.java:
```diff
@@ -58,4 +66,5 @@ public void clear() {
     head = null;
     tail = null;
   }
+
```
Let's not add this line.

pavibhai replied:
sure, removed this

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:
```diff
@@ -67,7 +70,7 @@
     Set<Integer> getColumnFilterIds();
 
-    Consumer<VectorizedRowBatch> getColumnFilterCallback();
+    Consumer<org.apache.orc.filter.FilterContext> getColumnFilterCallback();
```
Oh, we cannot import directly like `import org.apache.orc.filter.FilterContext`?

pavibhai replied:
This class also uses `org.apache.hadoop.hive.ql.io.filter.FilterContext` that is why this is fully qualified. I am happy to consider a different name e.g. `FilterInput`.

We can make this change after we have an approach on how to handle the breaking change of `Consumer<VectorizedRowBatch>` -> `Consumer<FilterContext>`

pavibhai replied:
Will change to OrcFilterContext

pavibhai replied:
Merged as part of ORC-755

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:
```diff
@@ -195,6 +198,8 @@ public boolean fileUsedProlepticGregorian() {
     protected BitFieldReader present = null;
     protected final Context context;
 
+    protected final ReadLevel rLevel;
```
`rLevel` -> `readLevel`?

pavibhai replied:
sure, changed

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:
```diff
@@ -316,12 +335,13 @@ protected long countNonNulls(long rows) throws IOException {
                            boolean[] isNull,
                            final int batchSize,
-                           FilterContext filterContext) throws IOException {
+                           FilterContext filterContext, ReadLevel rLevel) throws IOException {
```
In this case, shall we add this to the next line?

pavibhai replied:
sure, changed

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java:
```diff
@@ -144,10 +165,23 @@ public StripePlanner parseStripe(StripeInformation stripe,
+  }
+
+  public BufferChunkList readFollowData(OrcIndex index,
+                                          boolean[] rowGroupInclude,
```
indentation?

pavibhai replied:
done

dongjoon-hyun commented on java/core/src/findbugs/exclude.xml:
```diff
@@ -70,4 +70,8 @@
+  <Match>
+    <Bug pattern="URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD"/>
+    <Class name="org.apache.orc.impl.reader.StripePlanner$StreamInformation"/>
+  </Match>
```
Is this relevant to this PR? Can we proceed this independently?

pavibhai replied:
Removing this, not needed.

dongjoon-hyun commented on java/core/src/java/org/apache/orc/Reader.java:
```diff
@@ -382,7 +386,7 @@ public SearchArgument getSearchArgument() {
     }
 
-    public Consumer<VectorizedRowBatch> getFilterCallback() {
+    public Consumer<FilterContext> getFilterCallback() {
```
ditto.

dongjoon-hyun commented on java/core/src/java/org/apache/orc/filter/FilterContext.java:
```diff
@@ -0,0 +1,169 @@
+ * It offers a convenience method for finding the column vector from a given name, that the filters
+ * can invoke to get access to the column vector.
+ */
+public class FilterContext implements MutableFilterContext {
```
This is a simple wrapper for Hive's `VectorizedRowBatch` for easy access.
Can we propose this to Hive community?

pavibhai replied:
I think this is better suited in the ORC project as it is explicitly about ORC Filter handling.

pavibhai replied:
Merged as part of ORC-755

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/OrcAcidUtils.java:
```diff
@@ -68,7 +69,7 @@ public static long getLastFlushLength(FileSystem fs,
   }
 
-  private static final Charset utf8 = Charset.forName("UTF-8");
+  private static final Charset utf8 = StandardCharsets.UTF_8;
```
Please make another PR for this file change. We can merge that first.

pavibhai replied:
Merged as part of ORC-754

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/ReaderImpl.java:
```diff
@@ -74,7 +74,7 @@
   private final int metadataSize;
   protected final List<OrcProto.Type> types;
-  private TypeDescription schema;
+  private final TypeDescription schema;
```
Let's revert this change.

pavibhai replied:
Any reason for revert or you prefer this also goes into a separate PR?

dongjoon-hyun replied:
Yes, a separate PR, please.

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/RecordReaderUtils.java:
```diff
@@ -545,7 +545,7 @@ public int hashCode() {
 
     private long currentGeneration = 0;
 
-    private final TreeMap<Key, ByteBuffer> getBufferTree(boolean direct) {
```
Let's revert this change.

pavibhai replied:
Any reason for revert or you prefer this also goes into a separate PR?

pavibhai replied:
Merged as part of ORC-754

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:
```diff
@@ -663,41 +691,46 @@ public void checkEncoding(OrcProto.ColumnEncoding encoding) throws IOException {
-                           FilterContext filterContext) throws IOException {
+                           FilterContext filterContext,
+                           ReadLevel readLevel) throws IOException {
+
```
Please remove all empty line insertion.

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:
```diff
@@ -663,41 +691,46 @@ public void checkEncoding(OrcProto.ColumnEncoding encoding) throws IOException {
     @Override
-    public void skipRows(long items) throws IOException {
+    public void skipRows(long items, ReadLevel readLevel) throws IOException {
+
```
ditto. Please remove empty line insertion.

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/reader/tree/ReadLevel.java:
```diff
@@ -0,0 +1,25 @@
+public enum ReadLevel {
+  ALL,        // Read every column
+  LEAD,     // Read only the filter columns
+  FOLLOW   // Read the non-filter columns
```
Comment indentation is not aligned.

pavibhai replied:
fixed

dongjoon-hyun commented on java/core/src/java/org/apache/orc/impl/reader/tree/TypeReader.java:
```diff
@@ -28,18 +28,20 @@
                   boolean[] isNull,
                   int batchSize,
-                  FilterContext filterContext) throws IOException;
+                  FilterContext filterContext, ReadLevel readLevel) throws IOException;
```
This should be in a next line.

pavibhai replied:
done

dongjoon-hyun commented on site/develop/design/lazy_filter.md:
```diff
@@ -0,0 +1,352 @@
+* [Lazy Filter](#LazyFilter)
```
Thank you for adding this doc. Let's add this separately in order to focus the implementation.
Usually, the doc becomes obsolete during reviews. So, we had better focus on the document fully after code commit.

pavibhai replied:
Sure will pull this out into a separate PR

dongjoon-hyun commented on java/mapreduce/src/java/org/apache/orc/mapred/OrcMapredRecordReader.java:
```diff
@@ -98,16 +101,22 @@ public boolean next(NullWritable key, V value) throws IOException {
+                                            result.getFieldValue(i)));
+        } else {
+          result.setFieldValue(i, null);
+        }
```
Do you think we can proceed `OrcMapredRecordReader` and `OrcMapreduceRecordReader` separately first?

pavibhai replied:
Yes, we could. I need some guidance now on which changes can be combined or not.

So far we seem to have the following changes that are requested that we pull out separately:
1. Code changes like making a variable final or using the UTF8 Character Set
2. Changes to MapReduce classes
3. Findbugs exclude update

Can I create one PR that includes all of this or do you want separate PRs for each item that you requested a separate PR for?

dongjoon-hyun commented:
@pavibhai . FYI, Owen seems to change `master` massively soon in the following PR.
- https://github.com/apache/orc/pull/641

pavibhai commented:
> @pavibhai . FYI, Owen seems to change `master` massively soon in the following PR.
> 
>     * #641

Thanks for the heads up @dongjoon-hyun

pavibhai commented on site/develop/design/lazy_filter.md:
```diff
@@ -0,0 +1,352 @@
+* The performance of the filters is deteriorates with the increase of the IN values, however even in this case the
+  vector filter is much better than the row filter.
+
+In the next test we use a complex filter with both AND, and OR to understand the impact of Conjunctive Normal Form on
```
The existing SArg application only happens on File, Stripe and RG statistics so even in the worst case it will RowIndexStride times better than what we see with filter.

With the filter the evaluation happens on every row.

pavibhai commented on site/develop/design/lazy_filter.md:
````diff
@@ -0,0 +1,352 @@
+└──────────────────────────────────────────────────┘
+```
+
+The figure shows that we could read significantly fewer bytes by performing an additional read before reading to the end
````
In the example above the option was about allowing one partial read as a result that covered reading RG2 and then skips RG3 and RG4, starts at RG5 and reads to the end of the stripe.

If we configured allow 2 partial reads then it would not read RG6.

pavibhai replied:
No partial read configuration is not implemented.

pavibhai commented on java/core/src/java/org/apache/orc/filter/OrcFilterContext.java:
```diff
@@ -0,0 +1,169 @@
+  private static int findVectorIdx(String name, List<String> fieldNames) {
+    int idx = fieldNames.indexOf(name);
+    if (idx < 0) {
+      throw new IllegalArgumentException(String.format("Field %s not found in %s",
```
Here the focus is on the read schema. Schema evolution will already map the file schema into the read schema.
So in case the read schema has the column even if the file does not have the column the vector will be found and returned.
The case that is not supported is about find a column absent in the read schema.

pavibhai commented on java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:
```diff
@@ -83,17 +85,21 @@
   private long rowCountInStripe = 0;
   private final BatchReader reader;
   private final OrcIndex indexes;
+  private final boolean[] rowIndexCols;
```
will do

pavibhai replied:
sure, renamed

pavibhai commented on java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:
```diff
@@ -83,17 +85,21 @@
   private final int maxDiskRangeChunkLimit;
   private final StripePlanner planner;
+  private final ReadLevel readLevel;
+  private boolean needFollowStripe;
```
will do

omalley replied:
I think this should be inverted and name haveReadFollowData.

pavibhai replied:
renamed to needsFollowColumnsRead

pavibhai commented on java/core/src/java/org/apache/orc/impl/reader/tree/StructBatchReader.java:
```diff
@@ -15,62 +15,80 @@
-      this.context.getColumnFilterCallback().accept(batch);
+    if (readLevel != ReadLevel.FOLLOW) {
+      // In case of FOLLOW we leave the selectedInUse untouched.
+      batch.selectedInUse = false;
```
This is coming from the VectorizedRowBatch, it indicates a partial selection in the batch which requires the use of the selected vector to determine which rows are valid.

omalley commented:
With the update to master, I'm seeing a findbugs failure:

Dead store to readPercentage in org.apache.orc.TestRowFilteringIOSkip.filterAlternateBatches() [org.apache.orc.TestRowFilteringIOSkip] At TestRowFilteringIOSkip.java:[line 279] DLS_DEAD_LOCAL_STORE

pavibhai commented:
> With the update to master, I'm seeing a findbugs failure:
> 
> Dead store to readPercentage in org.apache.orc.TestRowFilteringIOSkip.filterAlternateBatches() [org.apache.orc.TestRowFilteringIOSkip] At TestRowFilteringIOSkip.java:[line 279] DLS_DEAD_LOCAL_STORE

thanks fixed this.

pgaref reviewed (commented):
First of all thanks @pavibhai for this PR introducing a neat feature! I have spent some time with row-filter and really appreciate this patch -- also left some initial comments.

I will have to agree with the existing comments though that this is just to hard to review and needs splitting. I would propose the following split:

- Findbugs onliner
- Changes to MapReduce classes
- Introduce OrcFilterContext (also affects row-filter tests etc. but is manageable)
- Introduce ReadLevel on Reader and Planner
- Code changes like making a variable final or using the UTF8 Character Set, or new metods (less important) 

Please let me know what you think

pavibhai commented:
> Left some initial comments here and there but in general I will have to agree with the existing comments that this is just to hard to review this PR and needs splitting. I would propose the following split:

Thanks for your feedback and suggestions @pgaref

> 
I would like to propose updates to checkstyle based on line and indentation feedback I have received on this PR.

>     * Findbugs onliner
Can you please clarify what you mean by this?
> 
>     * Changes to MapReduce classes
> 
>     * Introduce OrcFilterContext (also affects row-filter tests etc. but is manageable)
> 
>     * Introduce ReadLevel on Reader and Planner
> 
>     * Introduce [ORC-743](https://issues.apache.org/jira/browse/ORC-743) changes to the planner (partial-reads?)
[ORC-743](https://issues.apache.org/jira/browse/ORC-743) includes SArg to Filter conversion, the partial reads are part of this PR. The configuration for partial reads is absent in both the PRs right now.

I will treat this as introduce partial reads, followed by [ORC-743](https://issues.apache.org/jira/browse/ORC-743) which might have other splits that it needs.
> 
>     * Documentation changes after we merge all the expected functionality
> 
>     * Code changes like making a variable final or using the UTF8 Character Set, or new methods (less critical)
Any changes that are not required we can pull it out to an unrelated PR. I would like to include methods that avoid code duplication into the PR still.
> 
> 
> Please let me know what you think and again thanks for all the work!

Hope that makes sense. Please let me know.

pgaref commented:
> > Left some initial comments here and there but in general I will have to agree with the existing comments that this is just to hard to review this PR and needs splitting. I would propose the following split:
> 
> Thanks for your feedback and suggestions @pgaref
> 
> > 
> 
> I would like to propose updates to checkstyle based on line and indentation feedback I have received on this PR.
> 
> > ```
> > * Findbugs onliner
> > ```
> 
> Can you please clarify what you mean by this?
> 
> > ```
> > * Changes to MapReduce classes
> > 
> > * Introduce OrcFilterContext (also affects row-filter tests etc. but is manageable)
> > 
> > * Introduce ReadLevel on Reader and Planner
> > 
> > * Introduce [ORC-743](https://issues.apache.org/jira/browse/ORC-743) changes to the planner (partial-reads?)
> > ```
> 
> [ORC-743](https://issues.apache.org/jira/browse/ORC-743) includes SArg to Filter conversion, the partial reads are part of this PR. The configuration for partial reads is absent in both the PRs right now.
> 
> I will treat this as introduce partial reads, followed by [ORC-743](https://issues.apache.org/jira/browse/ORC-743) which might have other splits that it needs.
> 
> > ```
> > * Documentation changes after we merge all the expected functionality
> > 
> > * Code changes like making a variable final or using the UTF8 Character Set, or new methods (less critical)
> > ```
> 
> Any changes that are not required we can pull it out to an unrelated PR. I would like to include methods that avoid code duplication into the PR still.
> 
> > Please let me know what you think and again thanks for all the work!
> 
> Hope that makes sense. Please let me know.

> > * Findbugs onliner
What I meant: 
https://github.com/apache/orc/pull/635/files/4f400a3893aa7d7a32ffad7970041a6e8acf7c28#diff-e1aaa921b730eea0b914cb64c8adfe6e743a0395c114d60f26ee19e036da20e9

>I will treat this as introduce partial reads,
Sounds good to me -- lets just make it configurable

> I would like to include methods that avoid code duplication into the PR still.
Sure, it makes sense

@pavibhai Sounds like a plan -- lets start splitting this and I will be happy to help along the way!

pavibhai commented on java/core/src/test/org/apache/orc/TestRowFilteringIOSkip.java:
```diff
@@ -0,0 +1,415 @@
+public class TestRowFilteringIOSkip {
+  private final static Logger LOG = LoggerFactory.getLogger(TestRowFilteringIOSkip.class);
+  private static final Path workDir = new Path(System.getProperty("test.tmp.dir",
+                                                                  "target" + File.separator + "test"
```
I was aligning the parameters with indentation. 
Please let me know what is expected in terms of indentation?

pavibhai commented on java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:
```diff
@@ -2833,8 +2904,7 @@ public static TypeReader createTreeReader(TypeDescription readerType,
       case DECIMAL:
-        if (version == OrcFile.Version.UNSTABLE_PRE_2_0 &&
-            fileType.getPrecision() <= TypeDescription.MAX_DECIMAL64_PRECISION){
+        if (isDecimalAsLong(version, fileType.getPrecision())){
```
This is used in #636 I will remove from this PR and include it as part of that PR.

pavibhai commented on java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:
```diff
@@ -1221,52 +1244,112 @@ private boolean advanceToNextRow(
+        reader.setVectorColumnCount(batch.getDataColumnCount());
+        reader.nextBatch(batch, batchSize, readLevel);
+        if (readLevel == ReadLevel.LEAD && batch.size > 0) {
+          prepareFollowingStreams(rowInStripe, followRowInStripe);
```
added

pavibhai commented on java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:
```diff
@@ -748,11 +763,11 @@ private static TruthValue checkInBloomFilter(BloomFilter bf,
 
     if (predObj instanceof Long) {
-      if (bf.testLong(((Long) predObj).longValue())) {
+      if (bf.testLong((Long) predObj)) {
```
yeah not required for this patch, will move to separate PR

pavibhai replied:
Merged as part of ORC-754

pavibhai commented on java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:
```diff
@@ -1170,6 +1189,10 @@ private void advanceStripe() throws IOException {
     }
   }
 
+  private int computeRGIdx(long rowIdx) {
```
added

pavibhai commented on java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:
```diff
@@ -83,17 +85,24 @@
   private final int maxDiskRangeChunkLimit;
   private final StripePlanner planner;
+  // identifies the type of read: FULL (no filters), LEAD (filters are present)
+  private final ReadLevel readLevel;
```
FOLLOW is never set at the RecordReader level. It will only be FULL or LEAD. If this is LEAD then as part of read FOLLOW is implicitly used.

I can clarify further in the comment if that helps.

omalley replied:
I think I'd remove this and just check whether the row level filter is null or not.

pavibhai commented on java/core/src/java/org/apache/orc/impl/BufferChunkList.java:
```diff
@@ -22,6 +22,14 @@
  * Builds a list of buffer chunks
  */
 public class BufferChunkList {
+  public BufferChunk getHead() {
```
Removed

pavibhai commented on java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java:
```diff
@@ -363,11 +396,26 @@ private void addChunk(BufferChunkList list, StreamInformation stream,
     }
   }
 
+  private BufferChunkList addFollowChunk(long offset, long length) {
```
Removed

pavibhai commented on java/core/src/java/org/apache/orc/filter/OrcFilterContext.java:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  private ColumnVector findVector(String name) {
+    String[] refs = name.split(SPLIT_ON_PERIOD);
```
Maybe I am misinterpreting your comment. `RecordReaderImpl.findColumns` gives me id, but I need index to determine the column vector in the row batch.

I can use the ParserUtils.splitName instead of custom split.

pavibhai replied:
Merged as part of ORC-755

pavibhai commented on java/core/src/java/org/apache/orc/filter/OrcFilterContext.java:
```diff
@@ -0,0 +1,169 @@
+ * It offers a convenience method for finding the column vector from a given name, that the filters
+ * can invoke to get access to the column vector.
+ */
+public class OrcFilterContext implements MutableFilterContext {
```
will add to the documentation

pavibhai replied:
Merged as part of ORC-755

omalley commented on java/core/src/java/org/apache/orc/impl/reader/tree/ReadLevel.java:
```diff
@@ -0,0 +1,25 @@
+package org.apache.orc.impl.reader.tree;
+
+public enum ReadLevel {
+  ALL,     // Read every column
```
I think that we should remove ReadLevel.ALL and the functions like nextBatch should pass an EnumSet<ReadLevel>. Otherwise, I'm afraid that too many uses will fall through the cracks where ALL doesn't include the other values.

pavibhai replied:
good idea, will do

omalley commented on java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:
```diff
@@ -254,7 +273,7 @@ protected static IntegerReader createIntegerReader(OrcProto.ColumnEncoding.Kind
     }
 
-    public void startStripe(StripePlanner planner) throws IOException {
+    public void startStripe(StripePlanner planner, ReadLevel readLevel) throws IOException {
```
Do we really need the readLevel for start stripe? is startStripe(planner, FOLLOW) where we plan/read the next set of data?

pavibhai replied:
Yes, that is how currently we trigger [startStripe for FOLLOW](https://github.com/apache/orc/pull/635/files#diff-8678415090c906d11e0aa6241ba24643192cd66617c68d8deb0f7efe74dd2ba3R1311)

omalley commented on java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:
```diff
@@ -2854,11 +2924,19 @@ public static TypeReader createTreeReader(TypeDescription readerType,
+      return new StructBatchReader(reader, context);
+    } else if (reader instanceof LevelTypeReader
+               && ((LevelTypeReader)reader).getReader() instanceof StructTreeReader) {
+      return new LevelStructBatchReader((LevelTypeReader) reader, context);
```
I'm worried that doubling the number of TypeReaders in the tree is fair amount of overhead. Could we make it just at the transitions from LEAD to FOLLOW where we insert the extra LevelTypeReaders?

pavibhai replied:
I understand the concern regarding doubling the type readers. I don't quite understand the suggestion on how to circumvent that.

One possibility I do see is that instead of having the conditional code within LevelTypeReader we include that into each of the Readers directly, it will be duplicated but you will have a single reference.

pavibhai replied:
the logic has been moved into the Struct and Union TreeReaders instead.

omalley commented on java/core/src/java/org/apache/orc/impl/reader/tree/LevelTypeReader.java:
```diff
@@ -0,0 +1,107 @@
+                         int batchSize,
+                         FilterContext filterContext,
+                         ReadLevel readLevel) throws IOException {
+    if (reader.getReadLevel() != readLevel) {
```
Like I wrote elsewhere, I think this parameter should be an EnumSet<ReadLevel> and the test should be readLevel.contains(reader.getLevel()).

pavibhai replied:
will change

omalley commented on java/core/src/java/org/apache/orc/impl/reader/tree/LevelTypeReader.java:
```diff
@@ -0,0 +1,107 @@
+ * The methods are invoked on the reader only if the read level matches.
+ */
+public class LevelTypeReader implements TypeReader {
+  private final TypeReader reader;
```
Since the ReadLevel is only used here, can we take the ReadLevel field and getReadLevel() and move it from the TypeReader to here?

pavibhai replied:
LevelTypeReader has been removed.

pavibhai commented on java/core/src/java/org/apache/orc/Reader.java:
```diff
@@ -181,6 +183,8 @@
    * Options for creating a RecordReader.
    */
   class Options implements Cloneable {
+    private static final Logger LOG = LoggerFactory.getLogger(Options.class);
```
Will move this to #636

pavibhai commented:
@dongjoon-hyun @wgtmac @pgaref @omalley 
Thanks for your feedback and suggestions.

I will spend time on breaking this PR into [multiple PRs](https://github.com/apache/orc/pull/635#issuecomment-781490401) and submit.

Give me a couple of days for this change. I will try to incorporate all of the supplied feedback into them. If you have any other feedback feel free to post on this PR, I will still be tracking this PR or we can discuss further on the new PRs, whatever is convenient for you.

pavibhai commented on java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java:
```diff
@@ -245,7 +263,8 @@ protected static IntegerReader createIntegerReader(OrcProto.ColumnEncoding.Kind
         case DIRECT_V2:
         case DICTIONARY_V2:
-          return new RunLengthIntegerReaderV2(in, signed, context == null ? false : context.isSkipCorrupt());
+          return new RunLengthIntegerReaderV2(in, signed,
```
Merged as part of ORC-754

pavibhai commented on java/core/src/java/org/apache/orc/impl/reader/StripePlanner.java:
```diff
@@ -68,14 +74,15 @@
   private long currentStripeId;
   private long originalStripeId;
-  private Map<StreamName, StreamInformation> streams = new HashMap<>();
+  private final Map<StreamName, StreamInformation> streams = new HashMap<>();
```
Merged as part of ORC-754

pavibhai commented on java/core/src/java/org/apache/orc/filter/OrcFilterContext.java:
```diff
@@ -0,0 +1,169 @@
+ * It offers a convenience method for finding the column vector from a given name, that the filters
+ * can invoke to get access to the column vector.
+ */
+public class OrcFilterContext implements MutableFilterContext {
```
Merged as part of ORC-755

pavibhai commented on java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java:
```diff
@@ -234,11 +246,17 @@ protected RecordReaderImpl(ReaderImpl fileReader,
       LOG.info("Filter Columns: " + filterColIds);
+      this.readLevel = ReadLevel.LEAD;
+    } else {
+      this.readLevel = ReadLevel.ALL;
```
ReadLevel is now an EnumSet

pavibhai commented on java/core/src/java/org/apache/orc/impl/reader/tree/LevelStructBatchReader.java:
```diff
@@ -0,0 +1,28 @@
+
+import org.apache.orc.impl.TreeReaderFactory;
+
+public class LevelStructBatchReader extends StructBatchReader {
```
This class has been removed. the logic moves into Struct and Union TreeReaders instead.

pavibhai commented on java/core/src/java/org/apache/orc/impl/reader/tree/ReadLevel.java:
```diff
@@ -0,0 +1,25 @@
+
+package org.apache.orc.impl.reader.tree;
+
+public enum ReadLevel {
```
Pushing this into TypeReader as LevelTypeReader is no longer present.

pavibhai commented:
I have created a new Pull Request with the changes on LazyIO #668 and am closing this. I have addressed the comments previously called out in this PR into #668 except for ones that are not applicable as a result of the changes.

We can discuss this further in #668
